hpl-2.2/man/man3/HPL_equil.3
.TH HPL_equil 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_equil \- Equilibrate U and forward the column panel L.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&void\fR
\fB\&HPL_equil(\fR
\fB\&HPL_T_panel *\fR
\fI\&PBCST\fR,
\fB\&int *\fR
\fI\&IFLAG\fR,
\fB\&HPL_T_panel *\fR
\fI\&PANEL\fR,
\fB\&const enum HPL_TRANS\fR
\fI\&TRANS\fR,
\fB\&const int\fR
\fI\&N\fR,
\fB\&double *\fR
\fI\&U\fR,
\fB\&const int\fR
\fI\&LDU\fR,
\fB\&int *\fR
\fI\&IPLEN\fR,
\fB\&const int *\fR
\fI\&IPMAP\fR,
\fB\&const int *\fR
\fI\&IPMAPM1\fR,
\fB\&int *\fR
\fI\&IWORK\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_equil\fR
equilibrates  the  local  pieces  of U, so that on exit to
this function, pieces of U contained in every process row are of the
same size. This phase makes the rolling phase optimal.  In addition,
this  function probes  for  the  column panel L and forwards it when
possible.
.SH ARGUMENTS
.TP 8
PBCST   (local input/output)    HPL_T_panel *
On entry,  PBCST  points to the data structure containing the
panel (to be broadcast) information.
.TP 8
IFLAG   (local input/output)    int *
On entry, IFLAG  indicates  whether or not  the broadcast has
already been completed.  If not,  probing will occur, and the
outcome will be contained in IFLAG on exit.
.TP 8
PANEL   (local input/output)    HPL_T_panel *
On entry,  PANEL  points to the data structure containing the
panel (to be equilibrated) information.
.TP 8
TRANS   (global input)          const enum HPL_TRANS
On entry, TRANS specifies whether  U  is stored in transposed
or non-transposed form.
.TP 8
N       (local input)           const int
On entry, N  specifies the number of rows or columns of  U. N
must be at least 0.
.TP 8
U       (local input/output)    double *
On entry,  U  is an array of dimension (LDU,*) containing the
local pieces of U in each process row.
.TP 8
LDU     (local input)           const int
On entry, LDU specifies the local leading dimension of U. LDU
should be at least MAX(1,IPLEN[nprow]) when  U  is stored  in
non-transposed form, and MAX(1,N) otherwise.
.TP 8
IPLEN   (global input)          int *
On entry, IPLEN is an array of dimension NPROW+1.  This array
is such that IPLEN[i+1] - IPLEN[i] is the number of rows of U
in process IPMAP[i].
.TP 8
IPMAP   (global input)          const int *
On entry, IPMAP is an array of dimension  NPROW.  This  array
contains  the  logarithmic mapping of the processes. In other
words, IPMAP[myrow]  is the absolute coordinate of the sorted
process.
.TP 8
IPMAPM1 (global input)          const int *
On entry, IPMAPM1  is an array of dimension NPROW. This array
contains  the inverse of the logarithmic mapping contained in
IPMAP: For i in [0.. NPROCS) IPMAPM1[IPMAP[i]] = i.
.TP 8
IWORK   (workspace)             int *
On entry, IWORK is a workarray of dimension NPROW+1.
.SH SEE ALSO
.BR HPL_pdlaswp01N \ (3),
.BR HPL_pdlaswp01T \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/pgesv/HPL_equil.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#ifdef STDC_HEADERS
void HPL_equil
(
   HPL_T_panel *                    PBCST,
   int *                            IFLAG,
   HPL_T_panel *                    PANEL,
   const enum HPL_TRANS             TRANS,
   const int                        N,
   double *                         U,
   const int                        LDU,
   int *                            IPLEN,
   const int *                      IPMAP,
   const int *                      IPMAPM1,
   int *                            IWORK
)
#else
void HPL_equil
( PBCST, IFLAG, PANEL, TRANS, N, U, LDU, IPLEN, IPMAP, IPMAPM1, IWORK )
   HPL_T_panel *                    PBCST;
   int *                            IFLAG;
   HPL_T_panel *                    PANEL;
   const enum HPL_TRANS             TRANS;
   const int                        N;
   double *                         U;
   const int                        LDU;
   int *                            IPLEN;
   const int *                      IPMAP;
   const int *                      IPMAPM1;
   int *                            IWORK;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_equil equilibrates  the  local  pieces  of U, so that on exit to
 * this function, pieces of U contained in every process row are of the
 * same size. This phase makes the rolling phase optimal.  In addition,
 * this  function probes  for  the  column panel L and forwards it when
 * possible.
 *
 * Arguments
 * =========
 *
 * PBCST   (local input/output)          HPL_T_panel *
 *         On entry,  PBCST  points to the data structure containing the
 *         panel (to be broadcast) information.
 *
 * IFLAG   (local input/output)          int *
 *         On entry, IFLAG  indicates  whether or not  the broadcast has
 *         already been completed.  If not,  probing will occur, and the
 *         outcome will be contained in IFLAG on exit.
 *
 * PANEL   (local input/output)          HPL_T_panel *
 *         On entry,  PANEL  points to the data structure containing the
 *         panel (to be equilibrated) information.
 *
 * TRANS   (global input)                const enum HPL_TRANS
 *         On entry, TRANS specifies whether  U  is stored in transposed
 *         or non-transposed form.
 *
 * N       (local input)                 const int
 *         On entry, N  specifies the number of rows or columns of  U. N
 *         must be at least 0.
 *
 * U       (local input/output)          double *
 *         On entry,  U  is an array of dimension (LDU,*) containing the
 *         local pieces of U in each process row.
 *
 * LDU     (local input)                 const int
 *         On entry, LDU specifies the local leading dimension of U. LDU
 *         should be at least MAX(1,IPLEN[nprow]) when  U  is stored  in
 *         non-transposed form, and MAX(1,N) otherwise.
 *
 * IPLEN   (global input)                int *
 *         On entry, IPLEN is an array of dimension NPROW+1.  This array
 *         is such that IPLEN[i+1] - IPLEN[i] is the number of rows of U
 *         in process IPMAP[i].
 *
 * IPMAP   (global input)                const int *
 *         On entry, IPMAP is an array of dimension  NPROW.  This  array
 *         contains  the  logarithmic mapping of the processes. In other
 *         words, IPMAP[myrow]  is the absolute coordinate of the sorted
 *         process.
 *
 * IPMAPM1 (global input)                const int *
 *         On entry, IPMAPM1  is an array of dimension NPROW. This array
 *         contains  the inverse of the logarithmic mapping contained in
 *         IPMAP: For i in [0.. NPROCS) IPMAPM1[IPMAP[i]] = i.
 *
 * IWORK   (workspace)                   int *
 *         On entry, IWORK is a workarray of dimension NPROW+1.
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
   int                        i, ip, ipU, ipcur, iprow, iptgt, lastrow,
                              left, npm1, nprow, ll, llU, llcur, lltgt,
                              right, slen, smax, smin;
/* ..
 * .. Executable Statements ..
 */
   if( ( npm1 = ( nprow = PANEL->grid->nprow ) - 1 ) <= 1 ) return;
/*
 * If the current distribution of the pieces of U is already optimal for
 * the rolling phase, then return imediately.  The  optimal distribution
 * is such that ip processes have smax items and the remaining processes
 * only have smin items. Another way to check this is to verify that all
 * differences IPLEN[i+1] - IPLEN[i] are either smin or smax.
 */
   smax = ( ( slen = IPLEN[nprow] ) + npm1 ) / nprow;
   ip   = slen - nprow * ( smin = slen / nprow );

   iprow = 0;
   do
   {
      ll = IPLEN[iprow+1] - IPLEN[iprow]; iprow++;
   } while( ( iprow < nprow ) && ( ( ll == smin ) || ( ll == smax ) ) );

   if( iprow == nprow ) return;
/*
 * Now,  we are sure  the distribution of the pieces of U is not optimal
 * with respect to the rolling phase,  thus  perform  equilibration.  Go
 * through the list of processes:  Processes  that have rows that do not
 * belong to them  with respect to the optimal mapping spread them  in a
 * logarithmic fashion. To simplify a little bit the implementation, and
 * mainly the packing, a source process row spreads its data to its left
 * first, and then to its right.
 */
   IWORK[nprow] = slen;

   for( iprow = 0; iprow < nprow; iprow++ )
   {
      llU = IPLEN[iprow+1] - ( ipU = IPLEN[iprow] );
      if( iprow < ip ) { lltgt = smax; iptgt = iprow * smax;      }
      else             { lltgt = smin; iptgt = iprow * smin + ip; }

      left = ( ipU < iptgt ); right = ( iptgt + lltgt < ipU + llU );
/*
 * If I have something to spread to either the left or the right
 */
      if( ( llU > 0 ) && ( left || right ) )
      {        /* Figure out how much every other process should have */

         ipcur = ipU; llcur = llU;

         for( i = 0; i < nprow; i++ )
         {
            if( i < ip ) { lltgt = smax; iptgt = i * smax;      }
            else         { lltgt = smin; iptgt = i * smin + ip; }
            lastrow = iptgt + lltgt - 1;

            if( ( lastrow >= ipcur ) && ( llcur > 0 ) )
            { ll = lastrow - ipcur + 1; ll = Mmin( ll, llcur ); llcur -= ll; }
            else { ll = 0; }

            IWORK[i] = ipcur; ipcur += ll; IWORK[i+1] = ipcur;
         }
/*
 * Equilibration phase
 */
         if( TRANS == HplNoTrans )
         {
            if( left  )
            {
               HPL_spreadN( PBCST, IFLAG, PANEL, HplLeft,  N, U, LDU,
                            iprow, IWORK, IPMAP, IPMAPM1 );
            }

            if( right )
            {
               HPL_spreadN( PBCST, IFLAG, PANEL, HplRight, N, U, LDU,
                            iprow, IWORK, IPMAP, IPMAPM1 );
            }
         }
         else
         {
            if( left  )
            {
               HPL_spreadT( PBCST, IFLAG, PANEL, HplLeft,  N, U, LDU,
                            iprow, IWORK, IPMAP, IPMAPM1 );
            }

            if( right )
            {
               HPL_spreadT( PBCST, IFLAG, PANEL, HplRight, N, U, LDU,
                            iprow, IWORK, IPMAP, IPMAPM1 );
            }
         }
      }
   }
/*
 * Finally update  IPLEN  with the indexes corresponding to the new dis-
 * tribution of U - IPLEN[nprow] remained unchanged.
 */
   for( i = 0; i < nprow; i++ ) IPLEN[i] = ( i < ip ? i*smax : i*smin + ip );
/*
 * End of HPL_equil
 */
}
